### DISCUSSION TITLE
Workspaces vs namespaces

### DISCUSSION THREAD

**Discussion by vladimirvrabely** [astral-sh/rye#1162] [General]
Hi,
`rye` introduces _workspaces_ to manage multiple packages which seems to be inspired by cargo workspaces in Rust.
Interestingly, the provided workspace example uses a prefix for workspace members. I'm wondering what is the relation to [Python namespaces](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/)? 

The `rye` documentation doesn't seem to mention namespaces at all. Does it support them? Are workspaces meant to be an alternative to namespaces or are they complementary? Do you have any suggestions or opionions here? (I couldn't find any related issue or discussion)

**Comment by dynamicalsystem** [comment]
Yeah - you can jang it around with bash and it works fine.  There are some matters of taste to work out so I'm not committing this yet but here are a couple of scripts that do what you suggest.

### Assumptions

- I have a working rye installation
- I have read through the [Packaging namespace packages](https://packaging.python.org/en/latest/guides/packaging-namespace-packages/) doc
-  I buy into the [src-layout](https://packaging.python.org/en/latest/discussions/src-layout-vs-flat-layout/) view of the world
-  I don't mind the intermediary folder being called `namespace_package` (at the moment)
- I haven't yet published a package in this form and guess it will work

### Create a Namespace

I have a zsh script called `mknsp.sh` which:

1. Is used (on a mac) by calling `source mknsp.sh <namespace>`
2. Creates a `virtual` rye project
3. Turns that project into a workspace named for the namespace you want
4. `cd`s into the project directory
5. Copies another script for making packages/projects in the namespace project

```zsh
autoload colors; colors

rm -r "${1}"
rye init "${1}" --virtual
cd "${1}"

print -P "Adding…\n%F{green}[tool.rye.workspace]\nmembers = [\"${namespace}_*\"]%f"
print -P "to %F{cyan}./${1}/pyproject.toml%f"

echo "\n[tool.rye.workspace]\nmembers = [\"${1}_*\"]" >> ./pyproject.toml

cp ../mkpkg.sh .
```

### Create a Package

The other script, `mkpkg.sh`, does stuff:

1. Creates a rye project in a `namespace_package` sub directory of the `namespace` project
2. Moves the rye folders around to fit the namespace src-layout structure
3. Changes `[tool.hatch.build.targets.wheel]` to fit
4. Changes `[projects.scripts]` accordingly
5. Double checks the workspace is set up correctly

```zsh
autoload colors; colors

namespace=${PWD##*/}
subdir="${namespace}_${1}"

print -P "\nCreating the %F{cyan}${1}%f project in %F{cyan}…/${namespace}/${subdir}/%f"

rye init --script "${subdir}/"
mkdir "${subdir}/src/${namespace}"
mv "${subdir}/src/${subdir}" "${subdir}/src/${namespace}"
mv "${subdir}/src/${namespace}/${subdir}" "${subdir}/src/${namespace}/${1}"

tree --filesfirst

print -P "\nRemoving %F{red}_${1}%f from"
print -P "%F{green}[tool.hatch.build.targets.wheel]%f"
print -P "%F{green}packages = [\"src/${namespace}%f%F{red}_${1}%f%F{green}\"]%f"
print -P "in %F{cyan}${subdir}/pyproject.toml%f"

sed -i '' "/\[tool.hatch.build.targets.wheel\]/,/^\[/ s/src\/${namespace}_${1}/src\/${namespace}/" ${subdir}/pyproject.toml

print -P "%F{red}${namespace}_%f%F{green}${1} = \"${namespace}_${1}\"%f"
print -P "\nRemoving %F{red}${namespace}_%f from"
print -P "%F{green}[projects.scripts]%f"
print -P "in %F{cyan}${subdir}/pyproject.toml%f\n"

sed -i '' "/\[project.scripts\]/,/^/ s/\"${namespace}-${1}\"/${1}/" ${subdir}/pyproject.toml

print -P "\nTurning %F{green}${namespace}%f%F{red}_%f%F{green}${1}%f"
print -P "into %F{green}${namespace}%f%F{red}.%f%F{green}${1}%f"
print -P "in %F{green}[projects.scripts]%f"
print -P "in %F{cyan}${subdir}/pyproject.toml%f"

sed -i '' "/\[project.scripts\]/,/^/ s/_/./g" ${subdir}/pyproject.toml

if ! grep -q '^\[tool.rye.workspace\]' pyproject.toml; then
  print -P "Adding…\n%F{green}[tool.rye.workspace]\nmembers = [\"${namespace}_*\"]%f"
  print -P "to %F{cyan}./pyproject.toml%f"
  echo "\n[tool.rye.workspace]\nmembers = [\"${namespace}_*\"]" >> ../pyproject.toml
fi
```

### Example

```zsh
~ % source mknsp.sh namespace
```
```
success: Initialized virtual project in /Users/SimonHorrobin/Dev/namespace
  Run `rye sync` to get started
Adding…
[tool.rye.workspace]
members = ["namespace_*"]
to ./namespace/pyproject.toml
```

```zsh
namespace % source mkpkg.sh project
```
```
Creating the project project in …/namespace/namespace_project/
success: Initialized project in /Users/SimonHorrobin/Dev/namespace/namespace_project/
  Run `rye sync` to get started
.
├── README.md
├── mkpkg.sh
├── pyproject.toml
└── namespace_project
    ├── README.md
    ├── pyproject.toml
    └── src
        └── namespace
            └── project
                ├── __init__.py
                └── __main__.py

5 directories, 7 files

Removing _project from
[tool.hatch.build.targets.wheel]
packages = ["src/namespace_project"]
in namespace_project/pyproject.toml
namespace_project = "namespace_project"

Removing namespace_ from
[projects.scripts]
in namespace_project/pyproject.toml

Turning namespace_project
into namespace.project
in [projects.scripts]
in namespace_project/pyproject.toml
```
```zsh
namespace % source mkpkg.sh library
```
```
Creating the library project in …/namespace/namespace_library/
snipped
```
```zsh
namespace % rye sync
```
```
Reusing already existing virtualenv
Generating production lockfile: /Users/SimonHorrobin/Dev/namespace/requirements.lock
Generating dev lockfile: /Users/SimonHorrobin/Dev/namespace/requirements-dev.lock
Installing dependencies
Resolved 2 packages in 1ms
   Built namespace-library @ file:///Users/SimonHorrobin/Dev/namespace/namespace_lib
   Built namespace-project @ file:///Users/SimonHorrobin/Dev/namespace/namespace_pro
Prepared 2 packages in 622ms
Installed 2 packages in 1ms
 + namespace-library==0.1.0 (from file:///Users/SimonHorrobin/Dev/namespace/namespace_library)
 + namespace-project==0.1.0 (from file:///Users/SimonHorrobin/Dev/namespace/namespace_project)
Done!
```
```zsh
namespace % rye run project
```
`Hello from namespace-project!`
```zsh
namespace % rye run library
```
`Hello from namespace-library!`

### Usage

1. Add a `module.py` to `.../src/namespace/library`
2. Add a `def function():\n\treturn f"yo, {__name__}.function woz ere."` to `module.py`
3. Add `from namespace.library.module import function` to `.../src/namespace/project/__init__.py`
4. Add `\tprint(function())` to `namespace.project:main`
5. Run `namespace % rye run project
